src/Common/src/Microsoft/Win32/SafeHandles/NtlmSecuritySafeHandles.cs

```diff
@@ -0,0 +1,271 @@
+            byte[] target = new byte[length];
+            Interop.NetSecurity.ExtractNtlmBuffer(this, target, capacity: length, offset: offset);
+            return target;
+        }
+
+        public override bool IsInvalid
+        {
+            get { return handle == IntPtr.Zero; }
+        }
+
+        protected override bool ReleaseHandle()
+        {
+            Interop.NetSecurity.HeimNtlmFreeBuf(handle);
+            SetHandle(IntPtr.Zero);
+            return true;
+        }
+    }
+
+    /// <summary>
+    /// Wrapper around a session key used for signing
+    /// </summary>
+    internal sealed class SafeNtlmKeyHandle : SafeHandle
+    {
+        private uint _sequenceNumber;
+        private readonly bool _isSealingKey;
+        private readonly byte[] _digest;
+        private SafeEvpCipherCtxHandle _cipherContext;
+
+        // From MS_NLMP SIGNKEY at https://msdn.microsoft.com/en-us/library/cc236711.aspx
+        private const string s_keyMagic = "session key to {0}-to-{1} {2} key magic constant\0";
+        private const string s_client = "client";
+        private const string s_server = "server";
+        private const string s_signing = "signing";
+        private const string s_sealing = "sealing";
+
+        public SafeNtlmKeyHandle(byte[] key, bool isClient, bool isSealingKey)
+            : base(IntPtr.Zero, true)
+        {
+            string keyMagic = string.Format(s_keyMagic, isClient ? s_client : s_server,
+                    isClient ? s_server : s_client, isSealingKey ? s_sealing : s_signing);
+
+            using (IncrementalHash incremental = IncrementalHash.CreateHash(HashAlgorithmName.MD5))
+            {
+                 incremental.AppendData(key);
+                 incremental.AppendData(Encoding.UTF8.GetBytes(keyMagic));
+                 _digest = incremental.GetHashAndReset();
+            }
+
+            _isSealingKey = isSealingKey;
+            if (_isSealingKey)
+            {
+                _cipherContext = Interop.Crypto.EvpCipherCreate(Interop.Crypto.EvpRc4(), _digest, null, 1);
+            }
+        }
+
+        public override bool IsInvalid
+        {
+            get { return true; }
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing && (null != _cipherContext))
+            {
+                _cipherContext.Dispose();
+                _cipherContext = null;
+            }
+            base.Dispose(disposing);
+        }
+
+        protected override bool ReleaseHandle()
+        {
+            return true;
+        }
+
+        public byte[] Sign(SafeNtlmKeyHandle sealingKey, byte[] buffer, int offset, int count)
+        {
+            Debug.Assert(!_isSealingKey, "Cannot sign with sealing key");
+            Debug.Assert(offset >= 0 && offset < buffer.Length, "Cannot sign with invalid offset " + offset);
+            Debug.Assert((count + offset) <= buffer.Length, "Cannot sign with invalid count " + count);
```

Comment:
This assert is subject to overflow.  It needs to be written instead as `Debug.Assert(count <= buffer.Length - offset)`, which is safe because you've already proven that offset is between 0 and length.